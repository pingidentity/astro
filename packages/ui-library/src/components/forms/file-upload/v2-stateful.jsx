var React = require("react"),
    Constants = require("./v2-constants.js"),
    StatelessFileUpload = require("./v2-stateless"),
    Utils = require("../../../util/Utils"),
    fixOrientation = require("fix-orientation"),
    readExif = require("exif-js"),
    _ = require("underscore");

/**
 * @name FileUploadStateful
 * @memberof FileUpload
 * @desc This is a wrapper around the stateful (stateless=false) FileUpload to give the user image preview
 *    and error messages without having to implement that logic.
 */
module.exports = class extends React.Component {
    static displayName = "FileUploadStateful";

    static defaultProps = {
        onPreviewReady: _.noop
    };

    state = {
        errorMessage: "",
        thumbnailSrc: this.props.defaultImage || this.props.thumbnailSrc,
        labelAcceptedFileTypes: this.props.labelAcceptedFileTypes || this.props.accept.split("image/").join("")
    };

    _validateFileSize = (fileSizeInBytes) => {
        if (this.props.maxFileSizeKb && fileSizeInBytes > (this.props.maxFileSizeKb * 1000)) {
            this.resetComponent();
            throw Constants.ErrorCodes.TOO_BIG;
        }
    };

    _validateMimeType = (type) => {
        if (this.props.accept && !this.props.accept.match(new RegExp("\\b" + type + "\\b"))) {
            this.resetComponent();
            throw Constants.ErrorCodes.WRONG_MIME_TYPE;
        }
    };

    _fileReadSuccess = (file, contentUrl) => {
        var self = this;

        this.setState({ thumbnailSrc: contentUrl });
        this.props.onPreviewReady(contentUrl);

        if (contentUrl && file.type.match(/^image\/jpe?g$/i)) {
            //try to read exif data the base64 data was passed
            readExif.getData(file, function () {
                if (this.exifdata && this.exifdata.Orientation) {
                    //correct and display the image according to the exif orientation
                    fixOrientation(contentUrl, { image: true }, function (fixedImgContentUrl) {
                        self.setState({ thumbnailSrc: fixedImgContentUrl });
                        self.props.onPreviewReady(fixedImgContentUrl);
                    });
                }
            });
        }
    };

    _readFile = (file) => {
        var reader = new FileReader();

        reader.onloadend = function () { this._fileReadSuccess(file, reader.result); }.bind(this);
        reader.onerror = function () { this.props.onError(Constants.ErrorCodes.CANT_READ); }.bind(this);
        reader.readAsDataURL(file);
    };

    _processWithHtml5Api = (file, path) => {
        this.setState({
            fileName: Utils.stripFakePath(path)
        });

        try {
            this._validateFileSize(file.size);
            this._validateMimeType(file.type);

            if (this.props.showThumbnail && file.type.match("image.*")) {
                this._readFile(file);
            } else {
                this._processSimple();
            }
        } catch (err) {
            this.props.onError(err);
        }
    };

    _processSimple = () => {
        this.setState({ errorMessage: "", thumbnailSrc: "" });
    };

    _process = (files, path) => {
        if (!files) {
            return;
        }

        if (this.props.onValidate) {
            var errorMessage = this.props.onValidate(files[0]);

            if (errorMessage) {
                this.setState({ errorMessage: errorMessage });
                return;
            }
        }

        if (files && files[0]) {
            if (Utils.isHtmlFileApiSupported()) {
                this._processWithHtml5Api(files[0], path);
            } else {
                this._processSimple();
            }
        }
    };

    _handleChange = (e) => {
        if (this.props.onChange) {
            this.props.onChange(e);
        }
        this._process(e.target.files, e.target.value);
    };

    _handleRemove = () => {
        if (this.props.onRemove) {
            this.props.onRemove();
        }
        if (this.props.disabled) {
            return;
        }
        if (this.props.onValidate) {
            this.setState({
                errorMessage: this.props.onValidate()
            });
        }
        this.resetComponent();
    };

    /*
     * Reset the input
     */
    resetComponent = () => {
        this.setState({
            thumbnailSrc: "",
            fileName: ""
        });
        this.refs.FileUploadStateless.resetComponent();
    };

    render() {
        var props = _.defaults({
            ref: "FileUploadStateless",
            showFilename: !this.props.showThumbnail,
            onRemove: this._handleRemove,
            onChange: this._handleChange
        }, this.props, this.state);

        return <StatelessFileUpload {...props} />;
    }
};
