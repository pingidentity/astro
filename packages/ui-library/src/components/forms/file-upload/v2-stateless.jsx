import PropTypes from "prop-types";
import React, { Component } from "react";
import ReactDOM from "react-dom";
import classnames from "classnames";
import FormLabel from "../FormLabel";
import FormError from "../FormError";
import Button from "../../buttons/Button";
import { cannonballChangeWarning } from "../../../util/DeprecationUtils";
import { flagsPropType } from "../../../util/FlagUtils";

/**
 * @name FileUploadStateless
 * @memberof FileUpload
 * @desc This is a wrapper around the stateless (stateless=true) FileUpload.
 */
export default class extends Component {
    static displayName = "FileUploadStateless";

    static propTypes = {
        //labels
        labelText: PropTypes.string,
        label: PropTypes.string,
        labelSelect: PropTypes.string.isRequired,
        labelSelectOther: PropTypes.string,
        labelRemove: PropTypes.string.isRequired,
        labelMaxFileSize: PropTypes.string,
        //callbacks
        onRemove: PropTypes.func.isRequired,
        onChange: PropTypes.func.isRequired,
        //flags
        disabled: PropTypes.bool,
        showFilename: PropTypes.bool,
        showThumbnail: PropTypes.bool,
        showRemoveButton: PropTypes.bool,
        //properties
        defaultImage: PropTypes.string,
        errorMessage: PropTypes.string,
        filesAcceptedMessage: PropTypes.string,
        fileName: PropTypes.string,
        name: PropTypes.string,
        accept: PropTypes.string,
        "data-id": PropTypes.string,
        thumbnailSrc: PropTypes.string,

        flags: flagsPropType
    };

    _useTrueDefault = () => this.props.flags.includes("true-default")

    componentDidMount() {
        if (!this._useTrueDefault() && this.props.defaultImage) {
            cannonballChangeWarning({
                message: "The \"defaultImage\" parameter will no longer " +
                    "be the same as passing in thumbnailSrc. Instead, " +
                    "defaultImage will show an image when there is no " +
                    "current thumbnail, either due to not passing the prop " +
                    "or due to removing the image in the stateful version."
            });
        }
    }

    /*
     * Reset the input to the original state
     */
    resetComponent = () => {
        ReactDOM.findDOMNode(this.refs.fileInput).value = "";
    };

    render() {
        const fileSelected = !!(this.props.thumbnailSrc || this.props.fileName);
        const containerClass = classnames(this.props.className, "input-file-upload", {
            "image-upload": this.props.showThumbnail,
            "file-selected": fileSelected,
            disabled: this.props.disabled,
            "form-error": !!this.props.errorMessage
        });

        const {
            defaultImage,
            thumbnailSrc
        } = this.props;

        const useDefault = this._useTrueDefault() && defaultImage && !thumbnailSrc;

        return (
            <div className={containerClass} data-id={this.props["data-id"]}>
                <FormLabel
                    className={classnames({ "form-error": this.props.errorMessage }) }
                    value={this.props.labelText || this.props.label}>
                    <ImagePreview
                        isDefault={useDefault}
                        show={this.props.showThumbnail}
                        src={useDefault ? defaultImage : thumbnailSrc}
                    />
                    <input
                        disabled={this.props.disabled}
                        type="file"
                        ref="fileInput"
                        name={this.props.name || this.props.fileName}
                        accept={this.props.accept}
                        onChange={this.props.onChange}
                        data-id={this.props["data-id"] + "-input"}
                    />
                    <FileRestrictions
                        show={this.props.showThumbnail}
                        data-id={this.props["data-id"] + "-restrictions"}
                        labelMaxFileSize={this.props.labelMaxFileSize}
                        labelAcceptedFileTypes={this.props.labelAcceptedFileTypes}
                    />
                    <Button inline className="choose">
                        {(fileSelected && this.props.labelSelectOther) || this.props.labelSelect}
                    </Button>
                    {this.props.errorMessage && (
                        <FormError.Icon data-id={this.props["data-id"] + "-errormessage-icon"} />
                    )}
                    {this.props.errorMessage && (
                        <FormError.Message
                            value={this.props.errorMessage}
                            data-id={this.props["data-id"] + "-errormessage"}
                        />
                    )}
                </FormLabel>

                {fileSelected &&
                    <div className="file-info">
                        <Filename
                            show={this.props.showFilename}
                            data-id={this.props["data-id"] + "-fileName"}
                            value={this.props.fileName}
                        />
                        <a className="file-remove" ref="remove" onClick={this.props.onRemove}>
                            {this.props.labelRemove}
                        </a>
                    </div>
                }

                <AcceptMessage
                    data-id={this.props["data-id"] + "-filesAcceptedMessage"}
                    label={this.props.filesAcceptedMessage}
                />
            </div>
        );
    }
}

function ImagePreview({
    isDefault,
    src,
    show,
}) {
    return show ? (
        <div>
            {!isDefault && <span className="image-icon"></span>}
            <span
                className={classnames(
                    "input-image-thumb",
                    isDefault ? "input-image-thumb--default" : ""
                )}>
                <img src={src} data-id="imageThumb" alt="Thumbnail" />
            </span>
        </div>
    ) : null;
}

function FileRestrictions({
    "data-id": dataId,
    labelAcceptedFileTypes,
    labelMaxFileSize,
    show
}) {
    return show ? (
        <span className="file-size">
            <span className="max-size" data-id={dataId + "-max-size"}>
                {labelMaxFileSize}
            </span>
            <span className="accepted-types" data-id={dataId + "-accepted-types"}>
                {labelAcceptedFileTypes}
            </span>
        </span>
    ) : null;
}

function AcceptMessage({
    "data-id": dataId,
    value
}) {
    return value ? (
        <div className="image-types" data-id={dataId}>
            {value}
        </div>
    ) : null;
}

function Filename({
    "data-id": dataId,
    show,
    value
}) {
    return show ? (
        <span className="file-name" data-id={dataId}>
            <span className="icon-file"></span>
            {value}
        </span>
    ) : null;
}
