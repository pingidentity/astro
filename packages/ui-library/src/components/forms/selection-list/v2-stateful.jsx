var React = require("react"),
    _ = require("underscore"),
    Stateless = require("./v2-stateless.jsx"),
    filterItemsFunction = require("./v2-reducer").filterItemsFunction;

/**
 * @name SelectionListStateful
 * @memberof SelectionList
 * @desc This is a wrapper around the stateful (controlled=false) SelectionList to give the user search support
 *    without having to implement that logic.
 */
module.exports = React.createClass({
    displayName: "SelectionListStateful",

    getDefaultProps: function () {
        return {
            /*
             * The default search function does a "startsWith" search for query strings
             * less than or equal to 3 chars, otherwise is does a "contains" search.
             */
            onSearch: filterItemsFunction
        };
    },

    getInitialState: function () {
        return {
            queryString: "",
            matchedItems: this.props.items
        };
    },

    /**
     * @desc Filter items by search criteria.
     *     Items can be filtered by default search (3 first chars for start-search, 4 or more chars for contain-search)
     *     Items can be also filtered by a custom search implementation
     *
     * @param {string} queryString
     *     The new query string
     * @private
     */
    _onSearch: function (queryString) {
        var matchedItems = this.props.onSearch(this.props.items, queryString);
        this.setState({
            queryString: queryString,
            matchedItems: matchedItems
        });
    },

    /**
     * @desc toggle visibility of unselected items
     *
     * @private
     */
    _handleVisibilityChange: function () {
        this.setState({
            showOnlySelected: !this.state.showOnlySelected
        });
    },

    render: function () {
        var statelessProps = _.defaults(
            {
                ref: "SelectionListStateless",
                onSearch: this._onSearch,
                items: this.state.matchedItems,
                queryString: this.state.queryString,
                showOnlySelected: this.state.showOnlySelected,
                onVisibilityChange: this._handleVisibilityChange
            },
            this.props
        );

        return React.createElement(Stateless, statelessProps);
    }
});
