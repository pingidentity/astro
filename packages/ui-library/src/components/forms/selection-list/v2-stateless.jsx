import Button from "../../buttons/Button";
import PropTypes from "prop-types";
import React from "react";
import classnames from "classnames";
import _ from "underscore";
import If from "../../general/If";
import FlexRow, { justifyOptions } from "../../layout/FlexRow";
import FormRadioGroup from "../FormRadioGroup";
import FormCheckbox from "../FormCheckbox";
import FormSearchBox from "../FormSearchBox";
import HelpHint from "../../tooltips/HelpHint";
import { ListType } from "./v2-constants";
import { filterItemsFunction } from "./v2-reducer";

/**
 * @name SelectionListStateless
 * @memberof SelectionList
 * @desc This is a wrapper around the stateful (stateless=true) SelectionList.
 */
export default class SelectionListStateless extends React.Component {
    static displayName = "SelectionListStateless";

    static propTypes = {
        "data-id": PropTypes.string,
        className: PropTypes.string,
        type: PropTypes.oneOf([
            ListType.ADD,
            ListType.SINGLE,
            ListType.MULTI,
            ListType.MULTIADD,
            ListType.VIEWONLY
        ]),
        items: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.oneOfType([
                    PropTypes.number,
                    PropTypes.string
                ]).isRequired,
                name: PropTypes.string.isRequired
            })
        ),
        selectedItemIds: PropTypes.oneOfType([
            PropTypes.array,
            PropTypes.string,
            PropTypes.number
        ]),
        name: PropTypes.string,
        onValueChange: PropTypes.func,
        onSelectAll: PropTypes.func,
        showSearchBox: PropTypes.bool,
        searchPlaceholder: PropTypes.string,
        onSearch: PropTypes.func.isRequired,
        queryString: PropTypes.string,
        onVisibilityChange: PropTypes.func,
        showSelectionOptions: PropTypes.bool,
        showOnlySelected: PropTypes.bool,
        labelSelectAll: PropTypes.string,
        labelUnselectAll: PropTypes.string,
        labelOnlySelected: PropTypes.string,
        labelShowAll: PropTypes.string,
        optionsNote: PropTypes.node,
        requiredText: PropTypes.string,
        "no-border": PropTypes.bool,
        bottomPanel: PropTypes.node,
        multiAddButtonLabel: PropTypes.string,
        multiAddButtonDisabledHint: PropTypes.string,
        onMultiAdd: PropTypes.func,
        autoSelectAll: PropTypes.bool,
        autoFilter: PropTypes.bool,
    };

    static defaultProps = {
        "data-id": "selection-list",
        showSearchBox: true,
        type: ListType.SINGLE,
        showSelectionOptions: false,
        showOnlySelected: false,
        items: [],
        "no-border": false,
        multiAddButtonLabel: "Add",
        onMultiAdd: _.noop,
        onValueChange: _.noop,
        onSelectAll: _.noop,
        autoSelectAll: false,
        autoFilter: false,
    };

    /**
     * @desc Selects all checkboxes
     *
     * @private
     */
    _selectAll = () => {
        this.props.onSelectAll();
        if (this.props.autoSelectAll) {
            this.props.onValueChange(this._getItems().map(({ id }) => id));
        }
    };

    /**
     * @desc Toggle unchecking all checkboxes
     *
     * @private
     */
    _unselectAll = () => {
        this.props.onValueChange([]);
    };

    /**
     * @desc Toggle showing all or selected checkboxes
     *
     * @param {object} visibleItems currently displayed items on screen
     * @param {array} e the event object
     * @private
     */
    _onShowOnlyAllToggle = () => {
        this.props.onVisibilityChange();
    };

    /**
     * @desc Filter currently visible data on screen based on search string, toggles state, e.t.c.
     * @returns {array} the currently visible list of items
     * @private
     */
    _filterVisible = () => {
        return _.filter(this._getItems(), (item) => {
            return this.props.selectedItemIds.indexOf(item.id) > -1;
        });
    };

    _getSelectionOptions = () => {
        var itemsSelected = typeof(this.props.selectedItemIds) === "object"
            ? this.props.selectedItemIds.length : !!this.props.selectedItemIds;

        return (
            <div data-id={this.props["data-id"]} className="list-input__bottom-links">
                <a
                    data-id="show-only-or-all"
                    className="option"
                    onClick={this._onShowOnlyAllToggle}>
                    {this.props.showOnlySelected ? this.props.labelShowAll : this.props.labelOnlySelected}
                </a>
                {itemsSelected || !this.props.labelSelectAll
                    ? <a data-id="unselect-all" className="option" onClick={this._unselectAll}>
                        {this.props.labelUnselectAll}
                    </a>
                    : <a data-id="select-all" className="option" onClick={this._selectAll}>
                        {this.props.labelSelectAll}
                    </a>
                }
            </div>
        );
    };

    _renderMultiAddPanel = () => {
        const onClick = () => this.props.onMultiAdd(this.props.selectedItemIds);
        return (
            <FlexRow
                className="input-selection-list__multi-add-panel"
                justify={justifyOptions.CENTER}
            >
                <Button
                    data-id="add-button"
                    label={this.props.multiAddButtonLabel}
                    disabledText={this.props.multiAddButtonDisabledHint}
                    disabled={!!this.props.multiAddButtonDisabledHint}
                    onClick={onClick}
                    type="primary"
                />
            </FlexRow>
        );
    }

    // filter items if necessary
    _getItems = () => this.props.autoFilter
        ? filterItemsFunction(this.props.items, this.props.queryString)
        : this.props.items;

    render() {
        const {
            "data-id": dataId,
            optionsNote,
            requiredText,
            showSelectionOptions,
            type
        } = this.props;
        const className = classnames(
            this.props.className,
            "input-selection-list",
            {
                searchable: this.props.showSearchBox,
                "show-selection-options": showSelectionOptions,
                "input-selection-list--no-border": this.props["no-border"]
            });
        const visibleItems = this.props.showOnlySelected ? this._filterVisible() : this._getItems();

        return (
            <div data-id={dataId} className={className}>
                {requiredText && (
                    <div data-id={dataId + "-required-message"} className="required-message">
                        <span>{requiredText}</span>
                    </div>
                )}
                {this.props.showSearchBox && (
                    <div data-id={dataId + "-search-box"} className="selection-list-search">
                        <FormSearchBox
                            queryString={this.props.queryString}
                            placeholder={this.props.searchPlaceholder}
                            onValueChange={this.props.onSearch}
                            width="MAX"
                            flags={[ "p-stateful" ]}
                            {...this.props.searchBoxProps} // band-aid fix to allow overriding the stateful text field
                        />
                    </div>
                )}
                {optionsNote && <div className="input-selection-list__note">{optionsNote}</div>}
                <ListOptions
                    data-id={dataId + "-options"}
                    type={type}
                    selectedItemIds={this.props.selectedItemIds}
                    items={visibleItems}
                    onValueChange={this.props.onValueChange}
                    name={this.props.name}
                />
                <If test={showSelectionOptions}>
                    {this._getSelectionOptions(visibleItems)}
                </If>
                {
                    type === ListType.MULTIADD
                        ? this._renderMultiAddPanel()
                        : this.props.bottomPanel
                }
            </div>
        );
    }
}

/**
 * @class ListOptions
 * @desc SelectionList implements a list of selectable items with search capability.
 *
 * @param {string} [data-id="list-options"]
 *     To define the base "data-id" value for top-level HTML container
 * @param {string} [className]
 *     CSS classes to set on the top-level HTML container
 * @param {SelectionList.types} [type=SelectionList.types.SINGLE]
 *     Enum to specify the type of list items to render
 *         SINGLE for radio inputs next to each list item
 *         MULTI for checkbox inputs next to each list item
 *         VIEWONLY for text only list items
 * @param {SelectionListItem[]} items
 *     Actual data to display in the component
 * @param {array|string|number} [selectedItemIds]
 *     IDs of which list items are selected
 * @param {function} onValueChange
 *     Callback to be triggered when the item selection changes
 * @ignore
 */
class ListOptions extends React.Component {
    static propTypes = {
        "data-id": PropTypes.string,
        type: PropTypes.oneOf([
            ListType.ADD,
            ListType.SINGLE,
            ListType.MULTI,
            ListType.MULTIADD,
            ListType.VIEWONLY
        ]),
        items: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.oneOfType([
                    PropTypes.number,
                    PropTypes.string
                ]).isRequired,
                name: PropTypes.string.isRequired
            })
        ),
        selectedItemIds: PropTypes.oneOfType([
            PropTypes.array,
            PropTypes.string,
            PropTypes.number
        ]),
        onValueChange: PropTypes.func
    };

    static defaultProps = {
        "data-id": "list-options",
        type: ListType.SINGLE,
        onValueChange: _.noop
    };

    /**
    * @desc Generate list options as radio buttons
    * @return {object}
    *     The set of list options as radio buttons
    * @private
    * @ignore
    */
    _genAddOptions = () => {
        const valueChange = item => e => this.props.onValueChange(item, e);
        return this.props.items.map(item => (
            <div className="input-selection-list__add-option" key={item.id} onClick={valueChange(item)}>
                <Button
                    inline
                    iconName="plus"
                    data-id="row-button-add"
                    onClick={valueChange(item)}
                />
                {item.name}
            </div>
        ));
    }

    /**
    * @desc Generate list options as radio buttons
    * @return {object}
    *     The set of list options as radio buttons
    * @private
    * @ignore
    */
    _genRadioOptions = () => {
        return (
            <FormRadioGroup
                data-id={this.props["data-id"] + "-single-selection"}
                groupName={this.props.name || ("input-selection-list-items-" + this.props["data-id"])}
                items={this.props.items}
                stacked={true}
                selected={this.props.selectedItemIds}
                onValueChange={this.props.onValueChange}
            />
        );
    };

    /**
    * @desc Generate list options as checkboxes
    * @return {object}
    *     The set of list options as checkboxes
    * @private
    * @ignore
    */
    _genCheckboxOptions = () => {
        const isSelected = item => {
            return _.contains(this.props.selectedItemIds, item.id);
        };

        // add to the array of selected items (if it does not exist) or remove it (if it exists)
        const onSelectionValueChange = (item, checked) => () => {
            const updateFunction = checked ? _.union : _.difference;
            const updatedSelection = updateFunction(this.props.selectedItemIds, [item.id]);
            this.props.onValueChange(updatedSelection);
        };

        return this.props.items.map((item, index) => {
            const checked = isSelected(item);
            const onValueChangeFunc = onSelectionValueChange(item, !checked);

            return (
                <FormCheckbox
                    data-id={"selectionList-Checkbox-" + index}
                    key={item.id + "-" + index}
                    label={item.name}
                    className="stacked"
                    checked={checked}
                    onValueChange={onValueChangeFunc}
                    labelHelpText={item.helpHintText}
                    helpTarget={item.helpTarget}
                    name={this.props.name}
                />
            );
        });
    };

    /**
    * @desc Generate tooltip for item with helpHintText property
    * @param {object} item
    *     and object with the items properties including the helpHintText
    * @return {object}
    *     The help icon with tooltip text
    * @private
    * @ignore
    */
    _genTooltip = (item) => {
        return item.helpHintText
            ? <HelpHint
                hintText={item.helpHintText}
                placement="right"
                className="inline"
            />
            : null;
    };

    /**
    * @desc Generate list of view-only items
    * @return {object}
    *     The list of view-only options
    * @private
    * @ignore
    */
    _genViewonlyOptions = () => {
        return this.props.items.map((item, i) => {
            return (
                <div className="view-item" key={i}>
                    {item.name}{this._genTooltip(item)}
                </div>
            );
        });
    };

    /**
    * @desc Generate the set of list options based on the type
    * @return {object}
    *     The set of list options
    * @private
    * @ignore
    */
    _genListOptions = () => {
        switch (this.props.type) {
            case ListType.ADD:
                return this._genAddOptions();
            case ListType.SINGLE:
                return this._genRadioOptions();
            case ListType.MULTI:
            case ListType.MULTIADD:
                return this._genCheckboxOptions();
            case ListType.VIEWONLY:
                return this._genViewonlyOptions();
        }
    };

    render() {
        return (
            <div data-id={this.props["data-id"]} className="input-selection-list-items">
                {this._genListOptions()}
            </div>
        );
    }
}
