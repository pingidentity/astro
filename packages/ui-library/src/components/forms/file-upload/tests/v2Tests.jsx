window.__DEV__ = true;

jest.dontMock("./commonTests");
jest.dontMock("../index.js");
jest.dontMock("../v2");
jest.dontMock("../v2-stateful");
jest.dontMock("../v2-stateless");
jest.dontMock("../v2-constants.js");
jest.dontMock("../../FormError");
jest.dontMock("../../FormLabel");

//mock the exif api
jest.setMock("exif-js", { getData: jest.fn() });
jest.setMock("fix-orientation", jest.fn() );

describe("FileUpload", function () {
    const React = require("react");
    const ReactDOM = require("react-dom");
    const ReactTestUtils = require("react-dom/test-utils");
    const { mount } = require("enzyme");
    const TestUtils = require("../../../../testutil/TestUtils");
    const FileUpload = require("../index.js");
    const CommonTests = require("./commonTests");
    const Utils = require("../../../../util/Utils");
    const _ = require("underscore");

    //require these files so that jest will count them in coverage
    require("../v2-stateful");
    require("../v2-stateless");

    function getComponent (opts) {
        opts = _.defaults(opts || {}, {
            labelRemove: "Remove File",
            labelSelect: "Select File",
            labelMaxFileSize: "Max Size 4MB",

            "data-id": "testFileUpload",
            maxFileSizeKb: 10,
            showThumbnail: true,
            onValidate: jest.fn(),

            onPreviewReady: jest.fn(),
            onError: jest.fn()
        });

        return ReactTestUtils.renderIntoDocument(<FileUpload {...opts} />);
    }

    function get (component, part, opt) {
        var stateless = (component.refs.FileUploadStateful
            ? component.refs.FileUploadStateful
            : component).refs.FileUploadStateless;
        var findByDataId = TestUtils.findRenderedDOMNodeWithDataId.bind(TestUtils, component);

        switch (part) {
            case "input":
                return ReactDOM.findDOMNode(stateless.refs.fileInput);
            case "removeButton":
                return ReactDOM.findDOMNode(stateless.refs.remove);
            case "preview":
                return findByDataId("imageThumb");
            case "error-handler":
                return component.props.onError;
            case "error-message":
                return findByDataId(component.props["data-id"] + "-errormessage");
            case "accepted-types":
                return findByDataId(component.props["data-id"] + "-restrictions-accepted-types");
            case "max-size":
                return findByDataId(component.props["data-id"] + "-restrictions-max-size");
            case "filename":
                return findByDataId(component.props["data-id"] + "-fileName");
            case "files-accepted-message":
                return findByDataId(component.props["data-id"] + "-filesAcceptedMessage");
            case "enum":
                switch (opt) {
                    case "cantread":
                        return FileUpload.ErrorCodes.CANT_READ;
                    case "toobig":
                        return FileUpload.ErrorCodes.TOO_BIG;
                    case "wrongmime":
                        return FileUpload.ErrorCodes.WRONG_MIME_TYPE;
                }
        }
    }

    //Run the common tests
    CommonTests.run(getComponent, get);

    //bind the change simulator
    var simulateChange = CommonTests.simulateChange.bind(null, get);

    it("data-id's don't change", () => {
        TestUtils.mountSnapshotDataIds(
            <FileUpload />
        );
    });

    it("onchange callback", function () {
        var component = getComponent({ onChange: jest.fn() });

        simulateChange(component);

        expect(component.props.onChange).toBeCalled();
    });

    it("shows an error message when set", function () {
        var fileUploadErrorMsg = "Please upload a valid file",
            component = getComponent({ errorMessage: fileUploadErrorMsg }),
            error = get(component, "error-message");

        expect(error.textContent).toEqual(fileUploadErrorMsg);
    });

    it("does not show an error message when not set", function () {
        var component = getComponent(),
            error = get(component, "error-message");

        expect(error).toBeFalsy();
    });

    it("displays error returned by onValidate", function () {
        var component = getComponent({ onValidate: function () { return "this is my error"; } });

        simulateChange(component);

        var errorContainer = get(component, "error-message");

        expect(errorContainer).toBeTruthy();
        expect(errorContainer.textContent).toEqual("this is my error");
    });

    it("will generate an error when file size is more than the maximum size.", function () {
        var component = getComponent({ maxFileSizeKb: 10 });
        var errorHandler = get(component, "error-handler");

        simulateChange(component, { size: 21870 });

        expect(errorHandler).toBeCalledWith(get(component, "enum", "toobig"));
    });

    it("will set error text when file is not one of the accepted types.", function () {
        var component = getComponent();
        var errorHandler = get(component, "error-handler");

        simulateChange(component, { name: "someFile.bmp", type: "image/bmp" });

        expect(errorHandler).toBeCalledWith(get(component, "enum", "wrongmime"));
    });

    it("will reset without onValidate", function () {
        var component = getComponent({ onValidate: null, fileName: "file.jpg" });
        var fileInput = get(component, "input");
        var removeButton = get(component, "removeButton");

        ReactTestUtils.Simulate.click(removeButton);

        expect(fileInput.value).toEqual("");
    });

    it("will reset the value of the file input value when the remove link is clicked.", function () {
        const component = getComponent({ onValidate: jest.fn(), fileName: "file.jpg" });
        const removeButton = get(component, "removeButton");

        component.props.onValidate.mockClear();
        ReactTestUtils.Simulate.click(removeButton);

        // Jest no longer allows manually setting the value of inputs and calling onChange
        // on a file input requires creating a FileList (https://developer.mozilla.org/en-US/docs/Web/API/FileList)
        // There is no public API for creating one of these lists, so there isn't a good way to test this behavior.
        expect(component.props.onValidate).toBeCalled();
    });

    it("works when HTML5 Api is not available", function () {
        global.File = null;

        var component = getComponent();

        simulateChange(component);

        if (component.props.onFileChange) {
            expect(component.props.onFileChange).toBeCalled();
        } else {
            //the new component does not have an onFileChange callback
            expect(component.refs.FileUploadStateful.state.thumbnailSrc).toEqual("");
        }
    });

    it("throws error when deprecated prop 'controlled' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("controlled", "stateless"));

        expect(function () {
            getComponent({ controlled: true });
        }).toThrow(expectedError);
    });

    it("throws error when deprecated prop 'title' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("title", "labelText"));

        expect(function () {
            getComponent({ title: "something" });
        }).toThrow(expectedError);
    });

    it("true default: renders default image if no thumbnail is passed in", () => {
        const component = getComponent({
            defaultImage: "pictureofanuglydog.png",
            flags: ["true-default"]
        });

        const defaultImage = TestUtils.findRenderedDOMNodeWithClass(component, "input-image-thumb--default");

        expect(defaultImage).toBeTruthy();
    });

    it("true default: throws a warning when defaultImage is passed in without flag enabled", () => {
        console.warn = jest.fn();

        getComponent({
            defaultImage: "stuff.png"
        });

        expect(console.warn).toHaveBeenCalled();
    });

    it("true default: does not throw a warning when defaultImage is passed in with flag enabled", () => {
        console.warn = jest.fn();

        getComponent({
            defaultImage: "stuff.png",
            flags: ["true-default"]
        });

        expect(console.warn).not.toHaveBeenCalled();
    });

    it("fires cannonball warning for the stateless veraion", function() {
        console.warn = jest.fn();
        getComponent({ stateless: true });
        expect(console.warn).toBeCalled();
    });

    it("doesn't fire cannonball warning for the stateful veraion", function() {
        console.warn = jest.fn();
        getComponent({ stateless: false });
        expect(console.warn).not.toBeCalled();
    });

    it("renders node if given node as default image", () => {
        const component = mount(
            <FileUpload
                flags={["true-default"]}
                defaultImage={<div data-id="default-test" />}
            />
        );

        expect(component.find("[data-id=\"default-test\"]"));
    });

    it("Renders the file input button with a yellow border when required prop exists", () => {
        const component = mount(
            <FileUpload required/>
        );
        const element = component.find(".input-file-upload__select-btn--required");
        expect(element.exists()).toEqual(true);
    });

    it("Renders the file input button without yellow border when required prop does not exist", () => {
        const component = mount(
            <FileUpload />
        );
        const element = component.find(".input-file-upload__select-btn--required");
        expect(element.exists()).toEqual(false);
    });
});
