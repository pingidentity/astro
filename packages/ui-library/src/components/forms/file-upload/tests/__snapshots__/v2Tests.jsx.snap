// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload data-id's don't change 1`] = `
Array [
  "file-upload",
  "file-upload-input",
  "formLabel",
  "upload-button",
]
`;
